### ISSUE TITLE
Investigate libsodium bench results

### ISSUE THREAD

**Issue by codefromthecrypt** [tetratelabs/wazero#947]
https://github.com/jedisct1/libsodium has benchmarks which can be run by different backends

If you build the test wasm and run it like so, you can use wazero recently https://github.com/jedisct1/webassembly-benchmarks/issues/1#issuecomment-1019372687

```bash
$ WASI_RUNTIME=wazero ./dist-build/wasm32-wasi.sh --bench
```

Once these are built, it seems analogous to running like this:
```bash
$ wasmer test/default//shorthash.so
1580000
$ wasmtime run test/default//shorthash.so
1800000
$ wazero run test/default//shorthash.so
6730000
```

There's probably a good reason why our performance is slower especially as our existing benchmarks usually win. At first, I thought it was about pre-compilation until @mathetake helped me understand all the timing is done in-wasm, so our compilation won't matter except its end result. It would be good to get to the bottom of this and at least understand what's up.

@jedisct1 if you see any misunderstanding above, let us know.

**Comment by codefromthecrypt** [comment]
I hacked in our experimental function listener to see what was happening and one thing interesting is the clock choice is realtime (0) not monotonic (1). That seems an odd choice for benchmarks, but not necessarily an issue.

You can see the arg for clock id in `wasi_snapshot_preview1.clock_time_get(id=0,precision=1000,result.timestamp=1048536)`

Here's the whole dump
```
--> .$9()
	--> .$33()
		--> .$33()
		<--
		--> .$8()
		<--
	<--
	--> .$8()
	<--
	--> .$33()
	<--
	--> .$41(1052608,16)
		--> .$37(1052608,16)
			--> .$36(1052632,32)
				==> wasi_snapshot_preview1.random_get(buf=1052632,buf_len=32)
				<== ESUCCESS
			<-- 0
			--> .$38(1052664,1048464)
			<--
			--> .$38(1052664,1048464)
			<--
			--> .$38(1052728,1048464)
			<--
			--> .$38(1052792,1048464)
			<--
			--> .$38(1052856,1048464)
			<--
			--> .$38(1052920,1048464)
			<--
			--> .$38(1052984,1048464)
			<--
			--> .$38(1053048,1048464)
			<--
			--> .$38(1053112,1048464)
			<--
			--> .$10(1052608,1052664,16)
			<--
			--> .$11(1052664,0,16)
			<-- 1052664
		<--
	<--
	--> .$12(1048560)
		==> wasi_snapshot_preview1.clock_time_get(id=0,precision=1000,result.timestamp=1048536)
		<== ESUCCESS
	<-- 0
	--> .$12(1048560)
		==> wasi_snapshot_preview1.clock_time_get(id=0,precision=1000,result.timestamp=1048536)
		<== ESUCCESS
	<-- 0
	--> .$22(1051184,1050631,1048544)
		--> .$23(0,1050631,1048520,1048320,1048480)
		<-- 0
		--> .$18(1051184)
		<-- 0
		--> .$23(1051184,1050631,1048520,1048320,1048480)
			--> .$19(1050631,0,1051184)
				--> .$10(1051512,1050631,0)
				<--
			<--
			--> .$24(1040528,12,1048520)
			<--
			--> .$19(1050282,0,1051184)
				--> .$10(1051512,1050282,0)
				<--
			<--
			--> .$19(1040512,7,1051184)
				--> .$10(1051512,1040512,7)
				<--
			<--
			--> .$19(1050635,1,1051184)
				--> .$16(1051184,1050635,1)
					==> wasi_snapshot_preview1.fd_fdstat_get(fd=1,result.stat=1040056)
					<== ESUCCESS
					--> .$15(1051184,1050635,1)
						--> .$14(1,1040064,2)
							==> wasi_snapshot_preview1.fd_write(fd=1,iovs=1040064,iovs_len=2,result.nwritten=1040060)
6380000
							<== ESUCCESS
						<-- 8
					<-- 1
				<-- 1
				--> .$10(1051512,1050636,0)
				<--
			<--
		<-- 8
	<--
<--
```

**Comment by codefromthecrypt** [comment]
Adjusting the clock takes the time down, but not enough to be close to the others.

```bash
$ wazero run test/default//shorthash.so
5600000
```

**Comment by codefromthecrypt** [comment]
So, if someone wants to pick this up, basically the critical part would be between in function 9, between calls to function 12
[shorthash.so.gz](https://github.com/tetratelabs/wazero/files/10266189/shorthash.so.gz) is the wasm

**Comment by codefromthecrypt** [comment]
ps results above are from x86_64 macbook pro. if someone can run against arm, much obliged.

**Comment by jedisct1** [comment]
The entire set of wasm files used for the benchmark is here: https://github.com/jedisct1/webassembly-benchmarks/tree/master/2022-12/wasm

(but you already figured out how to reproduce them :) )

**Comment by codefromthecrypt** [comment]
another interesting thing is wasmer perf is more close to us on x86 if singlepass

```bash
$ wasmer run  test/default//shorthash.so
1620000

$ wasmer run --singlepass test/default//shorthash.so
5780000
```

**Comment by mathetake** [comment]
as at best our compiler backend is singlepass (eg no-global optimizations) plus in pure-Go JIT comes with a lot of constraints by the Go runtime (for example multiple registers are reserved by the Go runtime and  available to JIT, etc).

Anyway next year we are heading for optimizing compiler to improve the perfo https://github.com/tetratelabs/wazero/issues/777

**Comment by codefromthecrypt** [comment]
@jedisct1 so basically the results you had from last night seem legit based on the current arch of our go compiler engine. Doing dimensions for arm64 or other single-pass runtimes might be neat, but at least we know what's up. Cheers!

**Comment by codefromthecrypt** [comment]
one blindspot we can take action on is that our current benchmarks don't run a function in a loop inside wasm. So, folks running our benchmarks may not be aware that multi-pass backends will make that pattern faster and we won't (yet). I'll raise a PR to make this more transparent.

**Comment by mathetake** [comment]
one thing why amd64 is much slower vs arm64 is better that is we can't use `call` instruction at all (hence function calls are really heavy), but anyway enough detail on the Go runtime constraints

**Comment by codefromthecrypt** [comment]
SUMMARY:

wazero's current compiler backend won't win vs multi-pass where optimizable functions are called in a big loop (as well other optimizable code). End users can only optimize beforehand for now (ex `wasm-opt` or compiler flags). 

Different code will have different impact, and results will vary on different architectures. For example, in the benchmark we looked at arm64 would not be quite as slow, possibly percentage vs factors, due to its `call` instruction which we use.

Before seeing this, we had a blind spot because our "representative wasm" benchmarks wasn't representing a case of something called in a thousand or more scale loop in wasm. We fixed that blind spot in #948 so that when people run our benchmarks, they'll see we don't always win.

Many thanks to @jedisct1 for doing these benchmarks as in hindsight ours really weren't diverse enough and mostly focused on memory access and host functions, not crunching something in a loop.
